File: leetcode/implementation/1013. Partition Array Into Three Parts With Equal Sum/1013. Partition Array Into Three Parts With Equal Sum.cpp
class Solution {
public:
    bool canThreePartsEqualSum(vector<int>& arr) {
        int sum = 0;
        int pivSum = 0;
        for(auto a : arr) sum += a;
        if(sum % 3) return false;
        pivSum = sum / 3;
        sum = 0;
        vector <int> pivIdx;
        for(int i = 0; i < arr.size(); i++) {
            sum += arr[i];
            if(sum == pivSum) {
                pivIdx.push_back(i);
                sum = 0;
            }
        }
        return pivIdx.size() >= 3;
    }
};

File: leetcode/implementation/1013. Partition Array Into Three Parts With Equal Sum/1013. Partition Array Into Three Parts With Equal Sum.rs
use std::convert::TryInto;
impl Solution {
    pub fn can_three_parts_equal_sum(arr: Vec<i32>) -> bool {
        let mut sum: i32 = arr.iter().sum();
        if sum % 3 != 0 {
            return false;
        }
        let piv_sum = sum / 3;
        sum = 0;
        let mut piv_idx:Vec<i32> = Vec::new();
        for (i, &num) in arr.iter().enumerate() {
            sum += num;
            if sum == piv_sum {
                piv_idx.push(i.try_into().unwrap());
                sum = 0;
            }
        }
        piv_idx.len() >= 3
    }
}